File: source/ImplementQueueUsingStacks.cpp
// https://leetcode.com/problems/implement-queue-using-stacks
#include <stack>

class MyQueue
{
    std::stack<int> m_forwardStack;
    std::stack<int> m_reverseStack;
    int m_frontElement;
public:
    MyQueue() : m_forwardStack(), m_reverseStack(), m_frontElement(-1)
    {

    }

    void push(int x)
    {
        if (m_forwardStack.empty()) m_frontElement = x;
        m_forwardStack.push(x);
    }

    int pop()
    {
        while (!m_forwardStack.empty())
        {
            m_reverseStack.push(m_forwardStack.top());
            m_forwardStack.pop();
        }

        int result = m_reverseStack.top();
        m_reverseStack.pop();
        if (!m_reverseStack.empty()) m_frontElement = m_reverseStack.top();

        while (!m_reverseStack.empty())
        {
            m_forwardStack.push(m_reverseStack.top());
            m_reverseStack.pop();
        }

        return result;
    }

    int peek()
    {
        return m_frontElement;
    }

    bool empty()
    {
        return m_forwardStack.empty();
    }
};

File: CSharpPractice/source/ImplementQueueUsingStacks.cs
// https://leetcode.com/problems/implement-queue-using-stacks
using System.Collections.Generic;

namespace LeetCodePractice.ImplementQueueUsingStacks
{
    public class MyQueue
    {
        Stack<int> m_forwardStack = new Stack<int>();
        Stack<int> m_reverseStack = new Stack<int>();
        int m_frontElement = -1;

        public MyQueue()
        {
            
        }

        public void Push(int x)
        {
            if (m_forwardStack.Count == 0) m_frontElement = x;
            m_forwardStack.Push(x);
        }

        public int Pop()
        {
            while (m_forwardStack.Count > 0)
            {
                m_reverseStack.Push(m_forwardStack.Pop());
            }

            int result = m_reverseStack.Pop();
            if (m_reverseStack.Count > 0) m_frontElement = m_reverseStack.Peek();

            while (m_reverseStack.Count > 0)
            {
                m_forwardStack.Push(m_reverseStack.Pop());
            }

            return result;
        }

        public int Peek()
        {
            return m_frontElement;
        }

        public bool Empty()
        {
            return m_forwardStack.Count == 0;
        }
    }

    /**
     * Your MyQueue object will be instantiated and called as such:
     * MyQueue obj = new MyQueue();
     * obj.Push(x);
     * int param_2 = obj.Pop();
     * int param_3 = obj.Peek();
     * bool param_4 = obj.Empty();
     */
}